## Aim:
How to link the rule-based extracted CI_TYPE-GEO pairs with the respective Ci failure impacts

**Idea**\
Test using prompt engineering by passing table of CI-GEO pairs to GPT-J model.
Steps:
* Load model and apply it always on one chunk of the document to extract CI failure impacts 
* Use prompt engineering to extract time and location of the CI failure (origin) the CI impacts (impact location)
or 
* Pass dataframe of pairs as input to the model
or
* Use few shot prompting with example answers

**Finally:**
* Compaire all approaches of spatial and temporal linking CI failure impacts

In [1]:
# # # # In case of: ImportError: libGL.so.1: cannot open shared object file: No such file or directory
# # # ## NOTE opencv-python-headless needs to be version 4.13.0.92 and" langchain-docling==2.0.0"

# # RuntimeError: Module requires 'chunking' extra; to install, run: `pip install 'docling-core[chunking]'`

# !uv remove langchain-docling 
# !uv remove opencv-python 
# !uv remove opencv-python-headless
# !uv pip uninstall langchain-docling 
# !uv pip uninstall opencv-python 
# !uv pip uninstall opencv-python-headless
# !uv clean cache
# # # !uv add --no-cache-dir opencv-python-headless langchain-docling
# # !uv add "langchain-docling>=2.0.0"
# # # !uv add "opencv-python-headless==4.13.0.92" "langchain-docling>=2.0.0"
# !uv pip install --no-cache-dir opencv-python-headless
# # !uv pip install --no-cache-dir "opencv-python-headless==4.13.0.92" "langchain-docling==2.0.0"
# # # !uv pip install 'docling-core[chunking]'
# # # !uv add transformers
# !uv lock
# !uv sync --inexact

In [2]:
# #!source .venv/bin/activate
# # !uv pip install langchain-docling
# !uv pip install docling-core[chunking]
# !uv lock
# !uv sync --inexact
# from langchain_docling import DoclingLoader


In [3]:
# world_size = int(os.environ["WORLD_SIZE"])


In [4]:
import os

# # settings for CUDA and PYTORCH
os.environ["CUDA_DEVICE_ORDER"]="PCI_BUS_ID"
print(os.environ["CUDA_VISIBLE_DEVICES"])
os.environ["CUDA_VISIBLE_DEVICES"]="0,1"
os.environ["PYTORCH_ALLOC_CONF"]="expandable_segments:True" ## improve memory allocation
# # %env TORCH_CUDA_ARCH_LIST=8.6

# # settings for debugging CUDA errors (pinpoint exact line of error)
# os.environ["TORCH_USE_CUDA_DSA"] = "1"
# os.environ["CUDA_LAUNCH_BLOCKING"] = "1" 

# activate global venv explicitly
os.environ["VIRTUAL_ENV"] = "/beegfs/home/users/a/a-buch/_PROJECTS/CI-impacts-information-retrieval/.venv"


0,1


In [5]:
import torch
print(torch.cuda.is_available())
print(torch.cuda.device_count())  # should give 2
print(torch.cuda.get_device_name())
print(torch.cuda.get_device_properties(0))
# print(torch.cuda.get_device_properties(1))


True
2
NVIDIA A100-PCIE-40GB
_CudaDeviceProperties(name='NVIDIA A100-PCIE-40GB', major=8, minor=0, total_memory=40441MB, multi_processor_count=108, uuid=df26d912-ffe7-bf3e-615f-fecc33f8d120, pci_bus_id=130, pci_device_id=0, pci_domain_id=0, L2_cache_size=40MB)


In [6]:
# os.environ["WORLD_SIZE"]="1"
# os.environ["RANK"]="[0,1]"
# os.environ["LOCAL_RANK"]="[0,1]"
# # NOTE: # WORLD_SIZE: each GPU corresponds to one process (world = no. of processes within a group), processes communicate with each other enabling eg., distributed training
# # NOTE: # RANK: IDs of the processes, ranging from 0 up to WORLD_SIZE - 1


In [7]:
import os
import sys
import subprocess
import re
import time
from glob import glob
from pathlib import Path
import gc
from io import StringIO
import json

# from tqdm import tqdm
import numpy as np
import pandas as pd
# import pyarrow as pa
# import pyarrow.parquet as pq
import langdetect
from langchain_docling import DoclingLoader
import spacy
from huggingface_hub import login
from pdfminer.high_level import extract_text
from docling.backend.pypdfium2_backend import PyPdfiumDocumentBackend
from docling.datamodel.base_models import InputFormat
from docling.datamodel.pipeline_options import (
    PdfPipelineOptions,
    AcceleratorOptions,
    AcceleratorDevice,
)
from docling.pipeline.standard_pdf_pipeline import StandardPdfPipeline
from docling.document_converter import DocumentConverter, FormatOption


# sys.path.append("./")
from src.settings import settings as s
import src.document_cleaning as dc
import src.translation_model as tm
import src.extraction_model as em

from submodules.geollama.src.main import GeoLlama
from submodules.geollama.src.model import TopoModel, RAGModel

import transformers
import torch


test_mode = True

# login to HF
login(token=os.getenv("HUGGINGFACE_TOKEN")) 
# NOTE raises exception if not env.variable doesnt exist (compared to os.envrion.get and its shortcut os.getenv)


# NOTE. disabled batch size as OOM for CUDA despite chunkwise memory cleaning, nvtop to find best batchsize
BATCH_SIZE = s.BATCH_SIZE  # max for nvidia GPU

torch.manual_seed(42)

#  automatic linebreaks and multi-line cells.
pd.set_option('display.max_colwidth', 10000)
pd.set_option("display.colheader_justify", "left")

print(os.environ["CUDA_VISIBLE_DEVICES"])

# clean up before applying CUDA
gc.collect()
torch.cuda.empty_cache() 
print(torch.cuda.memory_reserved() / 1e9)
torch.no_grad()

# check if cuda cna be used
# device = transformers.infer_device()
# print(f"Using device: {device}")


# ## TODO test to prevent CUDA-OOM when reused
# ## Source: https://spacy.io/usage/embeddings-transformers
# from thinc.api import set_gpu_allocator, require_gpu

# # Use the GPU, with memory allocations directed via PyTorch.
# # This prevents out-of-memory errors that would otherwise occur from competing
# # memory pools.
# set_gpu_allocator("pytorch")
## require_gpu(0)



Running on TUB cluster
Running on TUB Cluster


/beegfs/home/users/a/a-buch/_PROJECTS/CI-impacts-information-retrieval/submodules/geollama/src/model.py:486: SyntaxWarning: invalid escape sequence '\o'
  return ['input ### Response: default output <\out>']
/beegfs/home/users/a/a-buch/_PROJECTS/CI-impacts-information-retrieval/submodules/geollama/src/model.py:503: SyntaxWarning: invalid escape sequence '\o'
  self.eos_token = '<\out>'


0
0.0


In [8]:
# !uv pip uninstall torch torchvision
# # !uv pip install https://github.com/mjun0812/flash-attention-prebuild-wheels/releases/download/v0.7.16/flash_attn-2.8.3+cu126torch2.8-cp310-cp310-linux_x86_64.whl
# # !uv pip install https://github.com/mjun0812/flash-attention-prebuild-wheels/releases/download/v0.9.4/flash_attn-2.8.3+cu126torch2.11-cp310-cp310-linux_x86_64.whl
# # !uv pip install https://github.com/mjun0812/flash-attention-prebuild-wheels/releases/download/v0.9.4/flash_attn-2.8.3+cu126torch2.11-cp312-cp312-linux_x86_64.whl
# !uv pip install torch torchvision --index-url https://download.pytorch.org/whl/cu126
# !uv lock
# # !uv sync
# # !uv sync --inexact  # not use otherwise overwritten
# !python -c "import torch; print(torch.cuda.get_arch_list())"

# # ## tested pytorch versions built against cuda 12.6 
# # ## issue to fix. ImportError: libcudnn.so.9: cannot open shared object file: No such file or directory
# # #  + torch==2.8.0
# # #  + torch==2.11.0+cu126 +  torchvision==0.26.0+cu126 -->uv snyc  torch==2.8.0 or torch==2.10


In [9]:
# !uv remove torch 
# !uv remove cudnn
# # # !uv remove transformers
# !uv pip uninstall torch 
# !uv pip uninstall cudnn 
# !uv add transformers 

# !uv cache clean
# !uv sync

# import os
# os.environ["LD_LIBRARY_PATH"] = "/beegfs/home/users/a/a-buch/_PROJECTS/CI-impacts-information-retrieval/.venv/lib/python3.10/site-packages/nvidia/cufile/lib"

# # !uv pip install cudnn
# # !uv pip install torch --index-url https://download.pytorch.org/whl/cu128
# #!uv pip install "torch<=2.8.0"
# !uv lock
# !uv sync  --inexact

import torch

import transformers


In [10]:
!uv version


ci-impacts-llm 0.1.0


In [11]:
# # # # !uv add "accelerate==0.33" "numpy==1.26.4"
# # !uv remove xformers # torch
# # #!uv pip uninstall xformers torch
# # # !uv cache clean
# # # # # !uv add unsloth
# !uv pip install --force-reinstall "xformers<0.0.26"

# # !uv add "bitsandbytes==0.42.0"




In [12]:
# # # !nvidia-smi
# # # !python -c "import torch; print(torch.cuda.is_available())"

# !source ./submodules/geollama/.venv/bin/activate
# # !source .venv/bin/activate

# # !uv remove xformers unsloth unsloth_zoo
# # !uv pip uninstall xformers unsloth unsloth_zoo
# # #pip install "unsloth[cu128-torch280] @ git+https://github.com/unslothai/unsloth.git"
# # !uv add "unsloth[cu128-torch280] @ git+https://github.com/unslothai/unsloth.git"
# #!uv add xformers


## FIXME install unsloth but cuda has to be 12.6 (due current HPC settings) + torch <= 2.9.2. (due to unsloth)

# # # # # get optimal downloads for unsloth 
# # # # # !wget -qO- https://raw.githubusercontent.com/unslothai/unsloth/main/unsloth/_auto_install.py | python -
# # # # !pip install --upgrade pip 
# # !uv pip install --no-deps git+https://github.com/unslothai/unsloth-zoo.git 
# !uv pip install "unsloth-zoo[cu126-torch211] @ git+https://github.com/unslothai/unsloth.git" --no-build-isolation
## not with --torch-backend=cu126  # as it changes torch >= 2.11 and torchvision as well
# # !uv pip install unsloth unsloth-zoo --torch-backend=cu126 
# !uv lock
# !uv sync


In [13]:
!lspci | grep -i vga

05:00.0 VGA compatible controller: ASPEED Technology, Inc. ASPEED Graphics Family (rev 30)


In [14]:
try: import torch
except: raise ImportError('Install torch via `pip install torch`')
from packaging.version import Version as V
import re
v = V(re.match(r"[0-9\.]{3,}", torch.__version__).group(0))
cuda = str(torch.version.cuda)
is_ampere = torch.cuda.get_device_capability()[0] >= 8
USE_ABI = torch._C._GLIBCXX_USE_CXX11_ABI
if cuda not in ("11.8", "12.1", "12.4", "12.6", "12.8", "13.0"): raise RuntimeError(f"CUDA = {cuda} not supported!")
if   v <= V('2.1.0'): raise RuntimeError(f"Torch = {v} too old!")
elif v <= V('2.1.1'): x = 'cu{}{}-torch211'
elif v <= V('2.1.2'): x = 'cu{}{}-torch212'
elif v  < V('2.3.0'): x = 'cu{}{}-torch220'
elif v  < V('2.4.0'): x = 'cu{}{}-torch230'
elif v  < V('2.5.0'): x = 'cu{}{}-torch240'
elif v  < V('2.5.1'): x = 'cu{}{}-torch250'
elif v <= V('2.5.1'): x = 'cu{}{}-torch251'
elif v  < V('2.7.0'): x = 'cu{}{}-torch260'
elif v  < V('2.7.9'): x = 'cu{}{}-torch270'
elif v  < V('2.8.0'): x = 'cu{}{}-torch271'
elif v  < V('2.8.9'): x = 'cu{}{}-torch280'
elif v  < V('2.9.1'): x = 'cu{}{}-torch290'
elif v  < V('2.9.2'): x = 'cu{}{}-torch291'
else: raise RuntimeError(f"Torch = {v} too new!")
if v > V('2.6.9') and cuda not in ("11.8", "12.6", "12.8", "13.0"): raise RuntimeError(f"CUDA = {cuda} not supported!")
x = x.format(cuda.replace(".", ""), "-ampere" if False else "") # is_ampere is broken due to flash-attn
print(f'pip install --upgrade pip && pip install --no-deps git+https://github.com/unslothai/unsloth-zoo.git && pip install "unsloth[{x}] @ git+https://github.com/unslothai/unsloth.git" --no-build-isolation')

pip install --upgrade pip && pip install --no-deps git+https://github.com/unslothai/unsloth-zoo.git && pip install "unsloth[cu128-torch280] @ git+https://github.com/unslothai/unsloth.git" --no-build-isolation


In [15]:
# !uv cache clean
# !uv pip install --upgrade --force-reinstall --no-cache-dir --no-deps unsloth unsloth_zoo
# !uv lock
!uv sync --no-exact

Resolved 302 packages in 24ms
Checked 289 packages in 453ms


## Set paths and vars

In [16]:
# set wd to project root
# os.chdir("/home/a-buch/Documents/TUB_DWN/_PROJECTS/CI-impacts-information-retrieval")

## set path variables
DOCS_DIR = Path(s.PATH_DATA / "text_sources/")
PARSED_TEXT_DIR = Path(s.PATH_DATA / "parsed_documents/")
NER_PATTERNS_FILEPATH = Path(s.NER_PATTERNS_FILEPATH)
LLM_OUTPUTS_DIR = Path(s.PATH_DATA / "llm_outputs/")
GEOLLM_OUTPUTS_DIR = Path(s.PATH_DATA / "geollm_outputs/")

os.makedirs(PARSED_TEXT_DIR, exist_ok=True)
os.makedirs(s.PATH_LLM_DATA, exist_ok=True)
os.makedirs(GEOLLM_OUTPUTS_DIR, exist_ok=True)


# CI GEO pairs
CI_GEO_FILEPATH = Path( s.PATH_DATA / s.CI_GEO_PAIRS_FILENAME)

## store LLM 1 response and prompt
OUTPUT_LLM1_FILEPATH =  Path(s.PATH_LLM_DATA / s.LLM_DATA_FILENAME)
OUTPUT_GEOLLM_FILEPATH =  Path(s.PATH_LLM_DATA / "geollm_results.csv")




### Set test mode

In [17]:

docs_list_sample = [
        # Path(PARSED_TEXT_DIR, "AEMET 2024 - ESTUDIO SOBRE LA SITUACIÓN DE LLUVIAS INTENSAS_cleaned.md"),
        # Path(PARSED_TEXT_DIR, "Karakatsani 2023 - Greece economy briefing The economic impact of the recent devastating floods in Greece_cleaned.md"),
        # Path(PARSED_TEXT_DIR, "Lloyd's List 2024 - Port of Valencia reopens after devastating floods_cleaned.md"),
        # Path(PARSED_TEXT_DIR, "Containerlift 2024 - Valencia Port Resumes Operations Following Devastating Flooding in Spain - Containerlift.co.uk - Transport_Lifting_Shipping_cleaned.md"), 
        #     Path(PARSED_TEXT_DIR, "ABC 2024 - Traffic jams and flight delays due to heavy rain and lightning storm in Malaga_cleaned.md"),

        Path(PARSED_TEXT_DIR, "Koks 2022 - Brief communication_cleaned.md"),
        # Path(PARSED_TEXT_DIR, "European Investment Bank 2025 - Spain_ EIB lends €50 million to Iberdrola to rebuild and climate-proof flood-hit power infrastructure in Valencia_cleaned.md"),
        # Path(PARSED_TEXT_DIR, "Wilson 2024 - Flash floods in Spain sweep away cars, disrupt trains and leave several missing _ AP News_cleaned.md"),     
        #     Path(PARSED_TEXT_DIR, "Wildhagen 2013 - Hochwasser_ Wie die Flut Unternehmen lahmlegt_cleaned.md"),

        # # # not Deidda et al, IPCC, Fekete 2025 as it already contains coarse info about many CI impacts
        #     # Path(PARSED_TEXT_DIR, "AFP 2022 - The_Vibes_Valencia Airport in Madrid briefly shut as lightning hits runway _ World _ The Vibes_cleaned.md"),
        #     # Path(PARSED_TEXT_DIR, "Artemis 2015 - PERILS finalises Storm Desmond UK flood loss estimate at £604m_cleaned.md"),
        #     # Path(PARSED_TEXT_DIR, "Brown 2010 - Economy feels chill as UK grinds to a halt _ The Independent_cleaned.md"),
        #     # Path(PARSED_TEXT_DIR, "Diakakis 2020 - A systematic assessment of the effects of extreme flash floods on transportation infrastructure and circulation: The example of the 2017 Mandra flood_cleaned.md"),
        # Path(PARSED_TEXT_DIR, "EFE 2024 - The DANA storm, live_ The death toll rises to 158_cleaned.md"),
        #     # Path(PARSED_TEXT_DIR, "Eurelectric 2006 - Impacts of Severe Storms on Electric Grids_cleaned.md"),
        #     # Path(PARSED_TEXT_DIR, "Euronews 2024 - Spain floods_ Death toll rises to 205 as nation braces for more rain _cleaned.md"),
        # Path(PARSED_TEXT_DIR, "Ferlita 2023 - Incendi in Sicilia, ecco cosa accade_cleaned.md"),
        #     # Path(PARSED_TEXT_DIR, "Fink 2004 - The 2003 European summer heatwaves and drought - synoptic diagnosis and impacts_cleaned.md"),
        # Path(PARSED_TEXT_DIR, "Gilbody Dickerson 2024 - Spain floods_ At least 95 people killed including British man near Malaga _ World News _ Sky News_cleaned.md"),
        #     # Path(PARSED_TEXT_DIR, "Kadir 2014 - The Impact of Natural Disasters on Critical Infrastructures - A Domino Effect-based Study_cleaned.md"),
        #     # Path(PARSED_TEXT_DIR, "Kaur 2025 - Authorities suspect arson in 17 wildfires across Dalmatian coast, Croatia - The Watchers_cleaned.md"),
        #     # Path(PARSED_TEXT_DIR, "Keller 2014 - Mapping Natural Hazard Impacts on Road Infrastructure—The Extreme Precipitation in Baden-Württemberg_cleaned.md"),
        #     # Path(PARSED_TEXT_DIR, "Kettle 2020 - Storm Xaver over Europe in December 2013 Overview of energy impacts and North Sea events_cleaned.md"),
        #     # Path(PARSED_TEXT_DIR, "Koks 2019 - Understanding Business Disruption and Economic Losses Due to Electricity Failures and Flooding_cleaned.md"),
        #     # Path(PARSED_TEXT_DIR, "Korzilius 2021 Nach der Flut_cleaned.md"),

        # # Path(PARSED_TEXT_DIR, "Khazai 2013 - Juni-Hochwasser 2013 in Mitteleuropa - Fokus Deutschland Bericht 2 Auswirkungen und Bewältigung_cleaned.md"),
            
        #     # not part of valid set:
        #     # Path(PARSED_TEXT_DIR, "Krausmann 2014 - STREST report on lessons learned from recent catastrophic events_cleaned.md"), # > 1800 entries LLMv3.0 incl. hallucinations
]


## Test mode
if test_mode:
    search_path = docs_list_sample
    print("Test mode is ON. Using only a small sample of documents for testing.")
else:
    search_path = glob(str(Path(PARSED_TEXT_DIR, "*cleaned.md")))

print(f"Using {len(search_path)} documents for processing.")

Test mode is ON. Using only a small sample of documents for testing.
Using 1 documents for processing.


In [13]:
# # install
# !uv remove transformers # [sentencepiece]
# !uv add transformers[sentencepiece]
# !uv add tokenizers tiktoken sentencepiece  #transformers[pretrained-tokenizer-fast]
# !uv lock
# !uv sync

###  Load spaCy language model

In [18]:
# !uv run python -m spacy download en_core_web_trf
# !uv pip install cmake
# !uv lock
# nlp = spacy.load("en_core_web_trf")


In [19]:
# # spaCy language model for NER and ENTITY LINKING

try: 
    print("Try loading spaCy language model for local machine ...")
    nlp = spacy.load(s.SPACY_MODEL)

except (OSError, ValueError):
    print(f"spaCy language model '{s.SPACY_MODEL}' not found. Downloading ...")
    subprocess.check_call(["uv", "pip", "install", "spacy-transformers"])
    subprocess.check_call(["uv", "run", "python3", "-m", "spacy", "download", s.SPACY_MODEL])
    nlp = spacy.load(s.SPACY_MODEL)
    # except ValueError:
    #     print("Try downloading directly via python3 -m spacy ... (and not via uv run ...)")
    #     subprocess.check_call(["python3", "-m", "spacy", "download", s.SPACY_MODEL])

# !uv pip install --upgrade unsloth unsloth_zoo
# !uv lock
# !uv sync
# !python -c "import torch; print(torch.cuda.is_available())"

# !uv run python -m spacy download en_core_web_trf
# nlp = spacy.load("en_core_web_trf")


#  - transformers==5.4.0 (from https://files.pythonhosted.org/packages/0b/a0/0a87883e564e364baab32adcacb4bec2e200b28a568423c8cf7fde316461/transformers-5.4.0-py3-none-any.whl)
#  - huggingface-hub==1.9.0


In [20]:
# Initialise spaCy pipeline 
nlp.add_pipe("merge_entities")
nlp.add_pipe("merge_noun_chunks")


# add CI_TYPE patterns to spacy nlp model pipeline
config = {"spans_key": None, "annotate_ents": True, "overwrite": False}
## see for more info: https://spacy.io/usage/rule-based-matching#entityruler
## NOTE EntityRuler is hidden inside .add_pipe()
try:
    ruler = nlp.add_pipe("span_ruler", config=config)
    ruler.from_disk(s.NER_PATTERNS_FILEPATH)
except ValueError:
    print("SpanRuler already exists in pipeline.")
    ruler = nlp.get_pipe("span_ruler")
    ruler.from_disk(s.NER_PATTERNS_FILEPATH)


## Document cleaning


In [14]:
# Docling pipeline configs

accelerator_options = AcceleratorOptions(
    num_threads=4, device=AcceleratorDevice.AUTO
)  # use GPU + multi-threading
pipeline_options = PdfPipelineOptions()
pipeline_options.do_ocr = True
pipeline_options.do_table_structure = (
    True  # identify tables as such just not to have them in the TextItems later
)
pipeline_options.accelerator_options = accelerator_options
pipeline_options.force_backend_text = True




# setup converter for PDF and markdown
converter = DocumentConverter(
    allowed_formats=[InputFormat.PDF, InputFormat.MD],
    format_options={
        InputFormat.PDF: FormatOption(
            pipeline_cls=StandardPdfPipeline,
            pipeline_options=pipeline_options,
            backend=PyPdfiumDocumentBackend,
        ),
    },
)


In [ ]:
print( "Number of documents to process:", len(os.listdir(DOCS_DIR)) )
start_time = time.time()


# convert the different layouts of the pdf files into unified markdown format incl. sub/section titles, tables, caption text etc
for pdf_filename in os.listdir(DOCS_DIR):

    if pdf_filename.endswith(".pdf"):

        md_filename = f"{Path(pdf_filename).stem}.md"

        pdf_filepath = os.path.join(DOCS_DIR, Path(pdf_filename))
        md_filepath = os.path.join(PARSED_TEXT_DIR, Path(md_filename))
        cleaned_md_filepath = md_filepath.replace(".md", "_cleaned.md")

        if os.path.exists(cleaned_md_filepath):
            print(f"Cleaned markdown file already exists: '{cleaned_md_filepath}'")
            continue

        print(f"\nFetching: {pdf_filename}")

        print("Remove reference section")
        pdf_text = extract_text(pdf_filepath)
        pdf_text_no_refs = dc.remove_references(pdf_text)

        print("Removing URLs") # LangExtract tries to open these URLs when they occur in the document text
        pdf_text_no_refs_urls = re.sub(r"http\S+", "", pdf_text_no_refs) 

        # FIXME remove workaround of saving pdf as markdown and reading it again as Docling.Document
        with open(md_filepath, "w", encoding="utf-8") as f:
            f.write(pdf_text_no_refs_urls)
        print("Converting Markdown to text...")
        # FIXME with DocLoader
        #loader = DoclingLoader(md_filepath)
        # md_text = loader.load()
        md_text = converter.convert(md_filepath)

        print("Removing headers and footers\n")
        md_text_cleaned = dc.remove_headers_footers(md_text)

        ## Translation
        src_language_doc = langdetect.detect(pdf_filename.replace(".pdf", "").lower())  # lower case improves language detection

        if src_language_doc != "en":
            supported_languages = ["fr", "de", "es", "it", "itc", "nl"]
            if src_language_doc not in supported_languages:
                print(f"Unsupported source language: {src_language_doc}. Continue with extraction on original text")
                continue 
            
            print(f"Translating {src_language_doc} --> en")
            md_text_cleaned.document.texts = tm.translate_2_english(src_language_doc, md_text_cleaned.document.texts)        


        print(f"Saving parsed and cleaned document as markdown to: {cleaned_md_filepath}")
        md_text_cleaned.document.save_as_markdown(cleaned_md_filepath)


end_time = time.time() - start_time
print(f"Parsing and cleaning done. Time elapsed: {end_time:.2f} seconds.")


# visual check of removed items
# TODO make as document_cleaning function: print removed items with largest number of chars first
# ## NOTE. high number of chars == more potentially actual text body

# text_items_removed = sorted(text_items_to_drop_visualization, key=lambda x: -x[0])
# for i in text_items_removed[:50]:
#     print(i) # -->  also subsection titles were removed partly





## LLama

###  Prompt engineering

In [18]:
question_1 = "Which infrastructure failures are mentioned in the text? Categorize the output by the type of infrastructure, the location, and the type of damage."

### Init Llama application function 
* extract CI-GEO pairs (ie. most likely geolocation of each CI_TYPE )
* pass NER table to prompt for evaluating and improving LLM response
* Test approach by applying it on three cleaned documents


In [19]:
!uv sync


Resolved 302 packages in 22ms
Checked 289 packages in 453ms


### Load GeoLlama

In [20]:
!source .venv/bin/activate
# # clean up after applying CUDA
import torch
from transformers import AutoTokenizer

import gc
# gc.collect() 
# torch.cuda.empty_cache() 
# torch.no_grad()
# print(torch.cuda.memory_reserved() / 1e9)



from submodules.geollama.src.main import GeoLlama
from submodules.geollama.src.model import TopoModel, RAGModel

import os
# os.environ["CUDA_VISIBLE_DEVICES"] = "0,1"
print(os.environ["CUDA_VISIBLE_DEVICES"])
# !nvidia-smi


0


In [21]:
# # !uv remove torch torchvision
# !uv pip uninstall torch torchvision
# !uv cache clean
# use torch==2.8.0
# # !uv pip install torch torchvision --default-index https://download.pytorch.org/whl/cu128 --force-reinstall 
# !uv pip install "transformers[sentencepiece]" tiktoken
# !uv lock
# !uv sync


# !uv add bitsandbytes
# !uv pip install -U bitsandbytes>=0.46.1




In [22]:
# # model_name = "JoeShingleton/GeoLlama-3.2-3b-toponym"
# model_name = "unsloth/Meta-Llama-3.1-8B-Instruct-bnb-4bit"

# tokenizer = AutoTokenizer.from_pretrained(
#     model_name,
#     #cache_dir="/beegfs/home/users/a/a-buch/_PROJECTS/CI-impacts-information-retrieval/notebooks/huggingface_mirror/hub", 
#     #use_fast=True  # use fast Rust-based tokenizer, when possible
# )

# # print("Model directory not found. Downloading model...")
# # # FastLanguageModel.for_inference(self.model)
# # print(f"FIXME - loading tokenizer as meta-llama/Llama-2-7b-chat-hf, due to backend issue, tokenizer needs sentencepiece/toiktoken, likely transformers need other version")
# # tokenizer = AutoTokenizer.from_pretrained(
# #     model_name, 
# #     # "meta-llama/Llama-2-7b-chat-hf",
# #     use_fast=True)

In [23]:
# !uv pip install --upgrade transformers
# !uv add tiktoken
# !uv lock
# !uv sync


In [24]:
# set up the model
topo_model = TopoModel(
    model_name='JoeShingleton/GeoLlama-3.2-3b-toponym',
    # model_name='JoeShingleton/GeoLlama_7b_toponym', 
    prompt_path='./submodules/geollama/data/prompt_templates/prompt_template.txt',
    instruct_path='./submodules/geollama/data/prompt_templates/topo_instruction.txt',
    input_path=None,
    config_path='./submodules/geollama/data/config_files/model_config.json'
)

rag_model = RAGModel(
    model_name='JoeShingleton/GeoLlama-3.2-3b-RAG',
    # model_name='JoeShingleton/GeoLlama_7b_RAG',
    prompt_path='./submodules/geollama/data/prompt_templates/prompt_template.txt',
    instruct_path='./submodules/geollama/data/prompt_templates/rag_instruction.txt',
    input_path='./submodules/geollama/data/prompt_templates/rag_input.txt',
    config_path='./submodules/geollama/data/config_files/model_config.json')


geo_llama = GeoLlama(
    topo_model = topo_model, 
    rag_model = rag_model, 
    translate_model=None
)

gc.collect()
torch.cuda.empty_cache() 
torch.no_grad()
print(torch.cuda.memory_reserved() / 1e9)



Using locally saved model


Loading weights:   0%|          | 0/254 [00:00<?, ?it/s]

Loading weights:   0%|          | 0/392 [00:00<?, ?it/s]

NOTE - loading tokenizer not from fine-tuned geollm model, but from its anchestor: unsloth/Meta-Llama-3.1-8B-Instruct-bnb-4bit
Using locally saved model


Loading weights:   0%|          | 0/254 [00:00<?, ?it/s]

Loading weights:   0%|          | 0/392 [00:00<?, ?it/s]

NOTE - loading tokenizer not from fine-tuned geollm model, but from its anchestor: unsloth/Meta-Llama-3.1-8B-Instruct-bnb-4bit
4.731174912


In [7]:
print(os.environ["CUDA_VISIBLE_DEVICES"])

# fake_input = "The briefing presents the economic impact of the damages caused by the storm Daniel in the region of Thessaly."
fake_input = "The  floods  caused  by  the  heavy  rainfall,  especially  in  Thessaly  plain -accounting  for approximately  15%  of  Greece’s  agricultural  land- resulted  to  a  massive  destruction  in agriculture,  infrastructure  and  residences,  which  is  expected  to  negatively  affect  the  Greek economy in short and mid-term. Fears for food shortages and increase of product prices have been raised. In addition, increase of fiscal costs, imports and unemployment may also affect the economy in the upcoming years. The extent of the impact to the economy is not yet know. Before the country recovered from the wildfires of August and the consequent huge forest disaster,  a  weather  stormy  phenomenon  named  Daniel  hit  the  country.  Specifically,  in  the beginning of September, rainfall of unprecedented intensity fell mainly in central Greece and especially  in  Thessaly  plain,  causing  floods  throughout  the  territory.  Thousands  of  acres  of crops and livestock farms were destroyed. Properties and houses were lost under ton of waters. National  roads  were  closed,  bridges  collapsed,  and  some  parts  of  the  railway  network  were highly damaged dividing the country in two."
resp_geollm = geo_llama.geoparse(fake_input)
resp_geollm

Both `max_new_tokens` (=512) and `max_length`(=131072) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


0,1
Error during Nominatim query: HTTPSConnectionPool(host='nominatim.openstreetmap.org', port=443): Read timed out. (read timeout=10)


[{'name': 'Thessaly',
  'latitude': 39.0,
  'longitude': 22.0,
  'RAG_estimated': False},
 {'name': 'Greece',
  'latitude': '38.9953683',
  'longitude': '21.9877132',
  'RAG_estimated': True}]

Checked 1 package in 252ms


### Apply Llama on chunks

In [25]:
## llamaattention as no apply_qkv

In [26]:
# !source .venv/bin/activate

# # # !uv remove xformers
# # # !uv remove flash-attn

# !uv pip uninstall torch 
# !uv pip uninstall transformers
# # # !uv pip uninstall setuptools
# # !uv pip uninstall torch
# # # !uv pip uninstall pytorch
# # !uv cache clean
# !uv pip install https://github.com/mjun0812/flash-attention-prebuild-wheels/releases/download/v0.7.16/flash_attn-2.8.3+cu128torch2.7-cp310-cp310-linux_x86_64.whl
# # # !uv add https://github.com/mjun0812/flash-attention-prebuild-wheels/releases/download/v0.7.2/flash_attn-2.8.3+cu128torch2.6-cp310-cp310-linux_x86_64.whl
# !uv pip install "torch<=2.8.0" #  torch==2.7.1
# !uv pip install transformers ipython
# # # !uv pip install torch torchvision --index-url https://download.pytorch.org/whl/cu128
# !uv lock
# !uv sync --inexact


In [10]:

gc.collect()
torch.cuda.empty_cache() 
torch.no_grad()
print(torch.cuda.memory_reserved() / 1e9)

0.0


In [ ]:
## Settings

model_name = "meta-llama/Llama-2-7b-chat-hf"
# TODO NOTE test for HPC: "meta-llama/Llama-3.1-8B-Instruct"

time0 = time.time()

## init LLM pipeline
decoder_model = em.DecoderModel(model_name)

## init output dataframe
df_responses_all = pd.DataFrame()
sentence_root_list = [] # for visualization of sentence roots and their dependencies
responses_error_list = []
df_geollm_response = pd.DataFrame()


if test_mode:
    search_path = docs_list_sample
    print("Test mode is ON. Using only a small sample of documents for testing.")
else:
    search_path = glob(str(Path(PARSED_TEXT_DIR, "*cleaned.md")))


## Start CI impact extraction
for i, filename in enumerate(search_path):

    src_language_nonengl = None 

    no_documents = len(search_path)
    filepath = Path(filename)
    filename_stem = filepath.stem



    ## load doc
    loader = DoclingLoader(filepath)  # use chunks from Docling.Loader
    doc = loader.load()


    print(f"\n\n ######## -------- Processing document [{i+1}/{no_documents}]: {filepath.name} -------- ######## \n")

    ## extract authors, publication year and title 
    author, year, title = dc.extract_citation_info(filename_stem)
    citation = f"{author} {year}".replace("  ", " ").strip()
    title = title.replace(" - ", "").replace("_cleaned", "").strip()



    print(f"\n ######## -------- Relationship extraction: CI-GEO pairs -------- ######## \n")

    ## Create New entity for transport infrastructure and apply it on any doc

    ## TODO make as pydantic class with fixed attributes
    df_ci_geo = pd.DataFrame(
        columns=[
            "citation_id",
            "chunk_id",
            "ci_entity",
            "geo_entity",
            "case_type",
            "chunk_text",
            "token_distance",
        ]
    )

    print(f"\n ######## -------- Getting geolocation of CI assets -------- ######## \n")
    for chunk_no, chunk in enumerate(doc):

        ## preprocess  TODO move to document cleaning workflow + dc.funcs
        chunk.page_content = chunk.page_content.replace("\n", " ")
        chunk.page_content = chunk.page_content.replace("- ", "-") # TODO test if ("- ", "") is better


    ## post-process of DF CI-GEO pairs
    print(
        "number of duplicates:", len(df_ci_geo.drop_duplicates(
            subset=["citation_id", "chunk_id", "ci_entity", "geo_entity","case_type", "chunk_text"]))
    )
    df_ci_geo = df_ci_geo.drop_duplicates(
        subset=["citation_id", "chunk_id", "ci_entity", "geo_entity","case_type", "chunk_text"]
        )# .reset_index(drop=True, inplace=True)



    print(f"\n  #############  -------- Text-2-Data: {filepath.name} -------- #############  \n")

    df_responses = pd.DataFrame(
        columns=[
            "citation_id",
            "chunk_id",
            "infrastructure_type",
            "damage",
            "location",
            "ci_entity",
            "geo_entity",
            "case_type",
            "chunk_text"
        ]
    )
    df_geollm_response = pd.DataFrame()

    
    ## apply decoder on each chunk in document
    ## TODO replace iteration by loading entire document and use recursive chunking from langchain
    for j, chunk in enumerate(doc):
        
        print("Chunk no", j)

        # clean up before applying CUDA
        gc.collect()
        torch.cuda.empty_cache() 
        torch.no_grad()
        # print(torch.cuda.memory_reserved() / 1e9)
        
        
        ## Start geoparsing with GeoLLM to verify Location
        # load GeoLLM prompt template
        template_geollm = em.load_prompt_template(template_filename="geollm.txt")
        
        
        # extract locations
        # for d in tqdm(chunk.page_content):
        resp = geo_llama.geoparse(chunk.page_content)
        # TODO add here geoLLm prompt as var
        # TODO check if useful in model.py to set : model.use_checkpointing = True or  model.gradient_checkpointing_enable()
        
        # save results
        df_geollm_resp = pd.DataFrame(resp[0:])
        df_geollm_resp["chunk_text"] = chunk.page_content
        
        df_geollm_response = pd.concat([df_geollm_response, df_geollm_resp], ignore_index=True)
        
        try:
            locations_per_chunk = df_geollm_resp["name"].to_list()
            lats_per_chunk = df_geollm_resp["latitude"].to_list()
            lons_per_chunk = df_geollm_resp["longitude"].to_list()
            RAGestimated_per_chunk = df_geollm_resp["RAG_estimated"].to_list()
        except:
            print("No locations extracted by GeoLLM for this chunk. Going to next chunk\n", resp[0:])
            pass
        
        print("NOTE -testing LLM without Entity linking")
        df_ci_geo = pd.DataFrame()
        # if df_ci_geo.loc[df_ci_geo["chunk_id"] == j].empty:
        
        print("df_geollm_resp resp0:]", resp[0:])
        context = [
            {
                "text": chunk.page_content,
                "citation": citation,
                "title": filename_stem, 
                "potential_locations": locations_per_chunk,  # list of 0 or multiple location strings
                
            },
        ]
        
        # load LLM_1 prompt template
        template = em.load_prompt_template(template_filename="ci_loc_direct_impacts_geollm.txt")
        
        # apply LLM 1
        response = decoder_model.generate_response(
            question=question_1, context=context, # chunk_id=j
            prompt_template = template,
            temperature = 0.2,
            max_new_tokens = 1024
        )
        
        ## postprocess response
        resp = response[0]["generated_text"].replace("\n", "")
        try:
            # fix missing bracket at beginning and end of response by splitting at first/last complete entry
            if "[{" not in resp:
                resp = "[{" + resp.split("{", 1)[1]
            if "}]" or "},]" not in resp:
                resp = resp.rpartition('}')[-3] + "}]"
        
            # remove potential text outside - no matter if it exists or not
            resp = ("[" + resp.split("[")[1]) 
            resp = (resp.split("]", 1)[0] + "]") 
        
            # save LLM response for each chunk  in dataframe for each document
            df_resp = pd.read_json(StringIO(resp))
            df_resp["citation_id"] = context[0]["citation"]  # add citation info
            df_resp["chunk_id"] = j  # add chunk id as identifier
            # df_resp["ci_entity"] = df_ci_geo.loc[df_ci_geo["chunk_id"] == j, "ci_entity"] 
            # df_resp["geo_entity"] = df_ci_geo.loc[df_ci_geo["chunk_id"] == j, "geo_entity"]
            # df_resp["case_type"] = df_ci_geo.loc[df_ci_geo["chunk_id"] == j, "case_type"]
            df_resp["chunk_text"] =  chunk.page_content #  FIXME context[0]["text"]  # add (translated) chunk text for tracing back LLM response
        
            df_responses = pd.concat([df_responses, df_resp], ignore_index=True)
        
        except (IndexError, ValueError) as e:
            print(f"Cannot add response: {e},\nFaulty response (before postprocessing):", response[0]['generated_text'].replace('\n', ''))
            # faulty response: e.g.  .., "location": "V" on satellite and online on radar"}, { ...}, {}
            responses_error_list.append({
                "citation_id": citation,
                "chunk_id": j,
                "response": response[0]['generated_text'].replace('\n', ''),
                "error": str(e)
            })
        
        
    df_responses_all = pd.concat([df_responses_all, df_responses], ignore_index=True)


    # clean up after each document
    gc.collect()
    torch.cuda.empty_cache()  # mainly needed after training, small effect when LLM applied only for inference
    torch.no_grad()

print(f"\n\n ######## -------- CI impact extraction took {(time.time() - time0) / 60} minutes -------- ######## \n\n")

/beegfs/home/users/a/a-buch/_PROJECTS/CI-impacts-information-retrieval/notebooks/huggingface_mirror/hub
Using locally saved model from /beegfs/home/users/a/a-buch/_PROJECTS/CI-impacts-information-retrieval/notebooks/huggingface_mirror/hub


Loading weights:   0%|          | 0/291 [00:00<?, ?it/s]

Test mode is ON. Using only a small sample of documents for testing.


Token indices sequence length is longer than the specified maximum sequence length for this model (584 > 512). Running this sequence through the model will result in indexing errors




 ######## -------- Processing document [1/1]: Koks 2022 - Brief communication_cleaned.md -------- ######## 


 ######## -------- Relationship extraction: CI-GEO pairs -------- ######## 


 ######## -------- Getting geolocation of CI assets -------- ######## 

number of duplicates: 0

  #############  -------- Text-2-Data: Koks 2022 - Brief communication_cleaned.md -------- #############  



Both `max_new_tokens` (=512) and `max_length`(=131072) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)
Both `max_new_tokens` (=1024) and `max_length`(=4096) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


Error parsing JSON: Unterminated string starting at: line 2 column 1574 (char 1574) 
- Raw response is  
{"toponyms": ["Netherlands", "Delft", "United Kingdom", "Irvine", "California", "United States of America", "Oxford", "Oxfordshire", "Oxfordshire", "Oxfordshire", "Irvine", "Irvine", "Oxfordshire", "Irvine", "Oxfordshire", "United Kingdom", "Oxfordshire", "Oxfordshire", "United Kingdom", "Oxfordshire", "Oxfordshire", "United Kingdom", "Oxfordshire", "Oxfordshire", "Oxfordshire", "Oxfordshire", "Oxfordshire", "United Kingdom", "Oxfordshire", "Oxfordshire", "Oxfordshire", "Oxfordshire", "Oxfordshire", "United Kingdom", "Oxfordshire", "Oxfordshire", "Oxfordshire", "Oxfordshire", "Oxfordshire", "Oxfordshire", "Oxfordshire", "Oxfordshire", "Oxfordshire", "Oxfordshire", "Oxfordshire", "Oxfordshire", "Oxfordshire", "Oxfordshire", "Oxfordshire", "Oxfordshire", "Oxfordshire", "Oxfordshire", "Oxfordshire", "Oxfordshire", "Oxfordshire", "Oxfordshire", "Oxfordshire", "Oxfordshire", "Oxfordshire

Both `max_new_tokens` (=512) and `max_length`(=131072) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)
Both `max_new_tokens` (=1024) and `max_length`(=4096) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


NOTE -testing LLM without Entity linking
df_geollm_resp resp0:] [{'name': 'Germany', 'latitude': '51.1638175', 'longitude': '10.4478313', 'RAG_estimated': True}, {'name': 'Netherlands', 'latitude': '52.2434979', 'longitude': '5.6343227', 'RAG_estimated': True}, {'name': 'Belgium', 'latitude': '50.6402809', 'longitude': '4.6667145', 'RAG_estimated': True}]


Both `max_new_tokens` (=512) and `max_length`(=131072) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


Cannot add response: list index out of range,
Faulty response (before postprocessing): 


Both `max_new_tokens` (=1024) and `max_length`(=4096) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


NOTE -testing LLM without Entity linking
df_geollm_resp resp0:] [{'name': 'Dutch', 'latitude': '51.2015196', 'longitude': '5.9046302', 'RAG_estimated': True}, {'name': 'German', 'latitude': 48.693, 'longitude': 11.235, 'RAG_estimated': False}, {'name': 'Belgian', 'latitude': '50.6402809', 'longitude': '4.6667145', 'RAG_estimated': True}, {'name': 'Rheinland-Pfalz', 'latitude': '49.9531599', 'longitude': '7.3106460', 'RAG_estimated': True}, {'name': 'Ahrtal', 'latitude': '50.3900652', 'longitude': '6.7335837', 'RAG_estimated': True}, {'name': 'Rheinland-Pfalz', 'latitude': '49.9531599', 'longitude': '7.3106460', 'RAG_estimated': True}, {'name': 'Rheinland-Pfalz', 'latitude': '49.9531599', 'longitude': '7.3106460', 'RAG_estimated': True}, {'name': 'Ahrtal', 'latitude': '50.3900652', 'longitude': '6.7335837', 'RAG_estimated': True}]


Both `max_new_tokens` (=512) and `max_length`(=131072) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)
Both `max_new_tokens` (=1024) and `max_length`(=4096) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


Error parsing JSON: Unterminated string starting at: line 2 column 1242 (char 1242) 
- Raw response is  
{"toponyms": ["Germany", "Vesdre River", "Liège", "Wavre", "Belgium", "Brussels", "Netherlands", "Trier", "Maaseik", "Herk-de-Stad", "Liège", "Ensival", "Halen", "Verviers", "Vesdre", "Belgium", "Pepinster", "Liège", "Verviers", "Ensival", "Verviers", "Belgium", "Belgium", "Herk-de-Stad", "Liège", "Verviers", "Verviers", "Verviers", "Belgium", "Liège", "Verviers", "Halen", "Liège", "Belgium", "Verviers", "Verviers", "Belgium", "Belgium", "Belgium", "Verviers", "Verviers", "Liège", "Verviers", "Verviers", "Verviers", "Belgium", "Verviers", "Verviers", "Verviers", "Verviers", "Verviers", "Verviers", "Verviers", "Verviers", "Verviers", "Verviers", "Verviers", "Verviers", "Verviers", "Verviers", "Verviers", "Verviers", "Verviers", "Verviers", "Verviers", "Verviers", "Verviers", "Verviers", "Verviers", "Verviers", "Verviers", "Verviers", "Verviers", "Verviers", "Verviers", "Verviers", "V

Both `max_new_tokens` (=512) and `max_length`(=131072) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)
Both `max_new_tokens` (=1024) and `max_length`(=4096) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


No locations extracted by GeoLLM for this chunk. Going to next chunk
 []
NOTE -testing LLM without Entity linking
df_geollm_resp resp0:] []


Both `max_new_tokens` (=512) and `max_length`(=131072) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)
Both `max_new_tokens` (=1024) and `max_length`(=4096) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


Error parsing JSON: Expecting value: line 2 column 2270 (char 2270) 
- Raw response is  
{"toponyms": ["western Europe", "Europe", "United Kingdom", "United States", "western Europe", "Europe", "United States", "western Europe", "western Europe", "Europe", "western Europe", "western Europe", "Europe", "western Europe", "western Europe", "western Europe", "western Europe", "western Europe", "western Europe", "western Europe", "western Europe", "Europe", "western Europe", "western Europe", "western Europe", "western Europe", "Europe", "western Europe", "western Europe", "western Europe", "Europe", "western Europe", "western Europe", "western Europe", "western Europe", "western Europe", "western Europe", "western Europe", "western Europe", "western Europe", "western Europe", "western Europe", "western Europe", "western Europe", "Europe", "western Europe", "western Europe", "western Europe", "western Europe", "western Europe", "western Europe", "western Europe", "western Europe", "western 

Both `max_new_tokens` (=512) and `max_length`(=131072) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)
Both `max_new_tokens` (=1024) and `max_length`(=4096) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


NOTE -testing LLM without Entity linking
df_geollm_resp resp0:] [{'name': 'Ahr valley', 'latitude': 49.7, 'longitude': 7.0, 'RAG_estimated': False}, {'name': 'Germany', 'latitude': '51.1638175', 'longitude': '10.4478313', 'RAG_estimated': True}]


Both `max_new_tokens` (=512) and `max_length`(=131072) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


Cannot add response: list index out of range,
Faulty response (before postprocessing): 


Both `max_new_tokens` (=1024) and `max_length`(=4096) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


NOTE -testing LLM without Entity linking
df_geollm_resp resp0:] [{'name': 'Ahr valley', 'latitude': 50.7, 'longitude': 6.8, 'RAG_estimated': False}, {'name': 'Belgium', 'latitude': '50.6402809', 'longitude': '4.6667145', 'RAG_estimated': True}, {'name': 'Pepinster', 'latitude': '50.5675217', 'longitude': '5.8037101', 'RAG_estimated': True}, {'name': 'Spa', 'latitude': '50.4729182', 'longitude': '5.8712335', 'RAG_estimated': True}, {'name': 'Netherlands', 'latitude': '52.2434979', 'longitude': '5.6343227', 'RAG_estimated': True}]


Both `max_new_tokens` (=512) and `max_length`(=131072) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


Cannot add response: list index out of range,
Faulty response (before postprocessing): 


Both `max_new_tokens` (=1024) and `max_length`(=4096) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


NOTE -testing LLM without Entity linking
df_geollm_resp resp0:] [{'name': 'Maastricht', 'latitude': '50.8512438', 'longitude': '5.6909768', 'RAG_estimated': True}, {'name': 'Liége', 'latitude': '50.4708135', 'longitude': '5.7735658', 'RAG_estimated': True}]


Both `max_new_tokens` (=512) and `max_length`(=131072) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


Cannot add response: list index out of range,
Faulty response (before postprocessing): 


Both `max_new_tokens` (=1024) and `max_length`(=4096) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


NOTE -testing LLM without Entity linking
df_geollm_resp resp0:] [{'name': 'Germany', 'latitude': '51.1638175', 'longitude': '10.4478313', 'RAG_estimated': True}, {'name': 'North Rhine-Westphalia', 'latitude': '51.4789205', 'longitude': '7.5543751', 'RAG_estimated': True}, {'name': 'Belgium', 'latitude': '50.6402809', 'longitude': '4.6667145', 'RAG_estimated': True}, {'name': 'Westnetz', 'latitude': '51.4978601', 'longitude': '7.4716388', 'RAG_estimated': True}, {'name': 'Ahr', 'latitude': '50.5125613', 'longitude': '6.9845010', 'RAG_estimated': True}, {'name': 'Belgium', 'latitude': '50.6402809', 'longitude': '4.6667145', 'RAG_estimated': True}, {'name': 'Rhineland-Palatinate', 'latitude': '49.9531599', 'longitude': '7.3106460', 'RAG_estimated': True}]


Both `max_new_tokens` (=512) and `max_length`(=131072) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)
You seem to be using the pipelines sequentially on GPU. In order to maximize efficiency please use a dataset
Both `max_new_tokens` (=1024) and `max_length`(=4096) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


NOTE -testing LLM without Entity linking
df_geollm_resp resp0:] [{'name': 'Belgium', 'latitude': '50.6402809', 'longitude': '4.6667145', 'RAG_estimated': True}, {'name': 'Netherlands', 'latitude': '52.2434979', 'longitude': '5.6343227', 'RAG_estimated': True}]


Both `max_new_tokens` (=512) and `max_length`(=131072) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


Cannot add response: list index out of range,
Faulty response (before postprocessing): 


Both `max_new_tokens` (=1024) and `max_length`(=4096) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


NOTE -testing LLM without Entity linking
df_geollm_resp resp0:] [{'name': 'Germany', 'latitude': '51.1638175', 'longitude': '10.4478313', 'RAG_estimated': True}, {'name': 'Rhineland-Palatinate', 'latitude': '49.9531599', 'longitude': '7.3106460', 'RAG_estimated': True}, {'name': 'Alte-nahr', 'latitude': 50.625, 'longitude': 6.625, 'RAG_estimated': False}, {'name': 'Mayschoss', 'latitude': '50.5220282', 'longitude': '7.0197776', 'RAG_estimated': True}, {'name': 'Sinzig', 'latitude': '50.5452042', 'longitude': '7.2490828', 'RAG_estimated': True}, {'name': 'GA', 'latitude': '32.3293809', 'longitude': '-83.1137366', 'RAG_estimated': True}, {'name': 'Erft', 'latitude': '51.1040645', 'longitude': '6.6138882', 'RAG_estimated': True}, {'name': 'Kuhn', 'latitude': '48.1008791', 'longitude': '11.5285431', 'RAG_estimated': True}]


Both `max_new_tokens` (=512) and `max_length`(=131072) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)
Both `max_new_tokens` (=1024) and `max_length`(=4096) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


NOTE -testing LLM without Entity linking
df_geollm_resp resp0:] [{'name': 'Germany', 'latitude': '51.1638175', 'longitude': '10.4478313', 'RAG_estimated': True}, {'name': 'Altenburg', 'latitude': '50.9308273', 'longitude': '16.8049522', 'RAG_estimated': True}, {'name': 'Heimersheim', 'latitude': '50.5425546', 'longitude': '7.1761142', 'RAG_estimated': True}, {'name': 'Ahr valley', 'latitude': 49.7, 'longitude': 6.8, 'RAG_estimated': False}, {'name': 'Germany', 'latitude': '51.1638175', 'longitude': '10.4478313', 'RAG_estimated': True}, {'name': 'Schuld', 'latitude': '50.4460848', 'longitude': '6.8868579', 'RAG_estimated': True}, {'name': 'Ahr valley', 'latitude': 49.7, 'longitude': 6.8, 'RAG_estimated': False}, {'name': 'Altenburg', 'latitude': '50.9308273', 'longitude': '16.8049522', 'RAG_estimated': True}, {'name': 'Ahr valley', 'latitude': 49.7, 'longitude': 6.8, 'RAG_estimated': False}]


Both `max_new_tokens` (=512) and `max_length`(=131072) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)
Both `max_new_tokens` (=1024) and `max_length`(=4096) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


NOTE -testing LLM without Entity linking
df_geollm_resp resp0:] [{'name': 'Bad Münstereifel', 'latitude': '50.5545674', 'longitude': '6.7636511', 'RAG_estimated': True}, {'name': 'Belgium', 'latitude': '50.6402809', 'longitude': '4.6667145', 'RAG_estimated': True}, {'name': 'Netherlands', 'latitude': '52.2434979', 'longitude': '5.6343227', 'RAG_estimated': True}]


Both `max_new_tokens` (=512) and `max_length`(=131072) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


Cannot add response: list index out of range,
Faulty response (before postprocessing): 


Both `max_new_tokens` (=1024) and `max_length`(=4096) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


NOTE -testing LLM without Entity linking
df_geollm_resp resp0:] [{'name': 'Ahrweiler', 'latitude': '50.5418300', 'longitude': '7.0951124', 'RAG_estimated': True}, {'name': 'Belgium', 'latitude': '50.6402809', 'longitude': '4.6667145', 'RAG_estimated': True}, {'name': 'Netherlands', 'latitude': '52.2434979', 'longitude': '5.6343227', 'RAG_estimated': True}, {'name': 'Rhineland-Palatinate', 'latitude': '49.9531599', 'longitude': '7.3106460', 'RAG_estimated': True}]


Both `max_new_tokens` (=512) and `max_length`(=131072) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)
Both `max_new_tokens` (=1024) and `max_length`(=4096) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


NOTE -testing LLM without Entity linking
df_geollm_resp resp0:] [{'name': 'Trier', 'latitude': '49.7596208', 'longitude': '6.6441878', 'RAG_estimated': True}, {'name': 'North Rhine-Westphalia', 'latitude': '51.4789205', 'longitude': '7.5543751', 'RAG_estimated': True}, {'name': 'Germany', 'latitude': '51.1638175', 'longitude': '10.4478313', 'RAG_estimated': True}, {'name': 'Rhineland-Palatinate', 'latitude': '49.9531599', 'longitude': '7.3106460', 'RAG_estimated': True}, {'name': 'Eschweiler', 'latitude': '50.5722896', 'longitude': '6.7404205', 'RAG_estimated': True}, {'name': 'Germany', 'latitude': '51.1638175', 'longitude': '10.4478313', 'RAG_estimated': True}, {'name': 'Eschweiler', 'latitude': '50.5722896', 'longitude': '6.7404205', 'RAG_estimated': True}, {'name': 'Germany', 'latitude': '51.1638175', 'longitude': '10.4478313', 'RAG_estimated': True}, {'name': 'Trier', 'latitude': '49.7596208', 'longitude': '6.6441878', 'RAG_estimated': True}, {'name': 'Germany', 'latitude': '51.16

Both `max_new_tokens` (=512) and `max_length`(=131072) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)
Both `max_new_tokens` (=1024) and `max_length`(=4096) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


NOTE -testing LLM without Entity linking
df_geollm_resp resp0:] [{'name': 'Germany', 'latitude': '51.1638175', 'longitude': '10.4478313', 'RAG_estimated': True}, {'name': 'Bad Neuenahr-Ahrweiler', 'latitude': '50.5444551', 'longitude': '7.1186671', 'RAG_estimated': True}, {'name': 'Belgium', 'latitude': '50.6402809', 'longitude': '4.6667145', 'RAG_estimated': True}, {'name': 'Netherlands', 'latitude': '52.2434979', 'longitude': '5.6343227', 'RAG_estimated': True}]


Both `max_new_tokens` (=512) and `max_length`(=131072) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)
Both `max_new_tokens` (=1024) and `max_length`(=4096) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


Error parsing JSON: Expecting value: line 2 column 1810 (char 1810) 
- Raw response is  
{"toponyms": ["Germany", "Europe", "Ahr valley", "Germany", "Germany", "Ahr", "Germany", "Ahr", "Germany", "Germany", "Germany", "Ahr", "Germany", "Ahr", "Germany", "Ahr", "Germany", "Germany", "Germany", "Ahr", "Germany", "Ahr", "Germany", "Germany", "Germany", "Germany", "Germany", "Germany", "Germany", "Germany", "Ahr", "Germany", "Germany", "Germany", "Germany", "Germany", "Germany", "Germany", "Germany", "Germany", "Germany", "Germany", "Germany", "Germany", "Germany", "Germany", "Germany", "Germany", "Germany", "Germany", "Germany", "Germany", "Germany", "Germany", "Germany", "Germany", "Germany", "Germany", "Germany", "Germany", "Germany", "Germany", "Germany", "Germany", "Germany", "Germany", "Germany", "Germany", "Germany", "Germany", "Germany", "Germany", "Germany", "Germany", "Germany", "Germany", "Germany", "Germany", "Germany", "Germany", "Germany", "Germany", "Germany", "Germany", "Ge

Both `max_new_tokens` (=512) and `max_length`(=131072) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


Cannot add response: list index out of range,
Faulty response (before postprocessing): 


Both `max_new_tokens` (=1024) and `max_length`(=4096) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


NOTE -testing LLM without Entity linking
df_geollm_resp resp0:] [{'name': 'Mayschoss', 'latitude': '50.5220282', 'longitude': '7.0197776', 'RAG_estimated': True}, {'name': 'Hönningen', 'latitude': '50.5157382', 'longitude': '7.3065922', 'RAG_estimated': True}, {'name': 'Ahr valley', 'latitude': 49.7, 'longitude': 6.9, 'RAG_estimated': False}]


Both `max_new_tokens` (=512) and `max_length`(=131072) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


Cannot add response: list index out of range,
Faulty response (before postprocessing): 


Both `max_new_tokens` (=1024) and `max_length`(=4096) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


NOTE -testing LLM without Entity linking
df_geollm_resp resp0:] [{'name': 'Germany', 'latitude': '51.1638175', 'longitude': '10.4478313', 'RAG_estimated': True}, {'name': 'Westnetz', 'latitude': '51.4978601', 'longitude': '7.4716388', 'RAG_estimated': True}, {'name': 'Belgium', 'latitude': '50.6402809', 'longitude': '4.6667145', 'RAG_estimated': True}, {'name': 'Netherlands', 'latitude': '52.2434979', 'longitude': '5.6343227', 'RAG_estimated': True}]


Both `max_new_tokens` (=512) and `max_length`(=131072) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


Cannot add response: list index out of range,
Faulty response (before postprocessing): 


Both `max_new_tokens` (=1024) and `max_length`(=4096) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


NOTE -testing LLM without Entity linking
df_geollm_resp resp0:] [{'name': 'Germany', 'latitude': '51.1638175', 'longitude': '10.4478313', 'RAG_estimated': True}]


KeyboardInterrupt: 

In [37]:
df_resp

,infrastructure_type,damage,location,citation_id,chunk_text
0,airport,severely damaged,Germany,Koks 2022,"Furthermore, in the region of Rhineland-Palatinate (Germany), 19 daycare centres and 17 schools suffered damage from the floods, affecting more than 8000 students (Staib, 2021). Approximately 4 months after the flood event, the district of Bad Neuenahr-Ahrweiler established emergency educational facilities using 297 con-tainers that serve as classrooms, offices and dining facilities for more than 800 students (Wiesbadener Kurier, 2021). In Belgium, various rural clinics have been affected and were unable to provide any services. Concurrently, in the most af-fected areas, general-practitioner facilities have been com-pletely destroyed (Le Spécialiste, 2021). In the Netherlands, one nursing home was flooded, and one hospital was evacu-ated as a precautionary measure."


### FIX. check geollm ~ llm_1 interaction

In [ ]:
chunk.page_content = "At the peak of the event, around 200 000 people experienced power outages in Germany. Electricity infrastructure was severely damaged in North Rhine-Westphalia and Rhineland-Palatinate. However, within 2 d around 50 % of the power was restored through repairs and temporary fixes. Within 8 weeks, no emergency power generators were required any-more, with most of the power infrastructure restored in Ger-many’s affected areas. Some areas, however, only had perma-nent power infrastructure after 6 months (Westnetz, 2022). The gas distribution network in the Ahr valley was severely damaged. Approximately 133 km of natural gas pipelines, 8500 gas metres, 3400 house pressure regulators, 7220 of the approximately 8000 household connections, and 31 systems measuring and regulating gas pressure have been damaged or destroyed (SWR, 2021). Gas supply was almost fully re-stored within 4.5 months after the flood event (Energienetze Mittelrhein, 2021). In Belgium, approximately 41 500 peo-ple experienced power outages at the peak of the event. This was the result of both damaged and deliberately switched-off electrical cabinets to prevent serious damages. It took around 3 weeks to fully restore power."
print(chunk.page_content)

df_responses = pd.DataFrame(
    columns=[
        "citation_id",
        "chunk_id",
        "infrastructure_type",
        "damage",
        "location",
        "ci_entity",
        "geo_entity",
        "case_type",
        "chunk_text"
    ]
)
df_geollm_response = pd.DataFrame()


# print("Chunk no", j)

# clean up before applying CUDA
gc.collect()
torch.cuda.empty_cache() 
torch.no_grad()
# print(torch.cuda.memory_reserved() / 1e9)

## Start geoparsing with GeoLLM to verify Location
# load GeoLLM prompt template
template_geollm = em.load_prompt_template(template_filename="geollm.txt")


# extract locations
# for d in tqdm(chunk.page_content):
resp = geo_llama.geoparse(chunk.page_content)
# TODO add here geoLLm prompt as var
# TODO check if useful in model.py to set : model.use_checkpointing = True or  model.gradient_checkpointing_enable()

# save results
df_geollm_resp = pd.DataFrame(resp[0:])
df_geollm_resp["chunk_text"] = chunk.page_content

df_geollm_response = pd.concat([df_geollm_response, df_geollm_resp], ignore_index=True)

locations_per_chunk = df_geollm_resp["name"].to_list()
lats_per_chunk = df_geollm_resp["latitude"].to_list()
lons_per_chunk = df_geollm_resp["longitude"].to_list()
RAGestimated_per_chunk = df_geollm_resp["RAG_estimated"].to_list()


print("NOTE -testing LLM without Entity linking")
df_ci_geo = pd.DataFrame()
# if df_ci_geo.loc[df_ci_geo["chunk_id"] == j].empty:

print("df_geollm_resp resp0:]", resp[0:])
context = [
    {
        "text": chunk.page_content,
        "citation": citation,
        "title": filename_stem, 
        "potential_locations": locations_per_chunk,  # list of 0 or multiple location strings
        
    },
]


############################################ SPACY
## preprocess  TODO move to document cleaning workflow + dc.funcs
chunk.page_content = chunk.page_content.replace("\n", " ")
chunk.page_content = chunk.page_content.replace("- ", "-") # TODO test if ("- ", "") is better

## get most likely geolocation for each CI entity based on distance between tokens
nlp_chunk = nlp(chunk.page_content)
all_ents = [ent for ent in nlp_chunk.ents]
ci_type_ents = [ent for ent in nlp_chunk.ents if ent.label_ in ["CI_TYPE", "FAC"]]


# check if chunk contains CI_TYPE entities
if len(ci_type_ents) > 0:

    # iterate over all entities within chunk
    for ent_idx in range(len(all_ents)):
        # when entity is CI_TYPE or FAC (i.e. buidling, airports, highways) do following ...
        if all_ents[ent_idx].label_ in ["CI_TYPE", "FAC"]:
            ci_idx = ent_idx

            ## .. calculate distances between CI_TYPE entity and  all GEO entities in chunk based on index position
            distance_list = []
            idx_in_chunk = []
            try:
                for ent_idx in range(len(all_ents)):
                    # TODO calc distances between CI_TYPE ~ GEO entities based on word numbers and not entities
                    if all_ents[ent_idx].label_ in ["GPE", "LOC"]:
                        geo_idx = ent_idx
                        dist_ent_pair = np.abs(ci_idx - geo_idx)
                        distance_list.append(dist_ent_pair)
                        idx_in_chunk.append((ent_idx))
                        closest_pair_idx = np.argmin(
                            distance_list
                        )  # idx of closest GEO entity
                        distance_closest_pair = distance_list[closest_pair_idx]

                threshold = 5  # max token distance between CI_TYPE and GEO entity
                if distance_closest_pair > threshold:
                    print(
                        f""" Token distance is too large between CI_TYPE/FAR and next GEO entity which is of {distance_closest_pair} [token distance] > {threshold} [max. token distance] """
                    )
                    continue
                else:
                    pass

                ## write as dict entry incl chunk_id, ci_entity, geo_entity, distance
                result_dict = {
                    "chunk_id": chunk_no,
                    "ci_entity": all_ents[ci_idx].text,
                    "ci_entity_label": all_ents[ci_idx].label_,
                    "geo_entity": all_ents[idx_in_chunk[closest_pair_idx]].text,
                    "geo_entity_label": all_ents[idx_in_chunk[closest_pair_idx]].label_,
                    "token_distance": distance_closest_pair,
                }
                df_ci_geo = pd.concat(
                    [df_ci_geo, pd.DataFrame([result_dict])], ignore_index=True
                )

            except IndexError:
                continue
else:
    # print("\nNo CI_TYPE or FAC entities found in this chunk.")
    continue


## post-process of DF CI-GEO pairs
print(
    "number of duplicates:", len(df_ci_geo.drop_duplicates(
        subset=["citation_id", "chunk_id", "ci_entity", "geo_entity","case_type", "chunk_text"]))
)
df_ci_geo = df_ci_geo.drop_duplicates(
    subset=["citation_id", "chunk_id", "ci_entity", "geo_entity","case_type", "chunk_text"]
    )# .reset_index(drop=True, inplace=True)

###########################
...
##########################

]
# else:  # TODO dissolve if else clause by making it in ci_locations: if df_ci_geo.chunk=j, xx, else None
#     context = [
#                 {
#                     "text": chunk.page_content,
#                     "citation": citation,
#                     "title": filename_stem, 
#                     "ci_locations": df_ci_geo.loc[df_ci_geo["chunk_id"] == j],
#                 },
#             ]

########################################### LLM 1

# load LLM_1 prompt template
template = em.load_prompt_template(template_filename="ci_loc_direct_impacts_geollm.txt")

# apply LLM 1
response = decoder_model.generate_response(
    question=question_1, context=context, # chunk_id=j
    prompt_template = template,
    temperature = 0.2,
    max_new_tokens = 1024
)

## postprocess response
resp = response[0]["generated_text"].replace("\n", "")
try:
    # fix missing bracket at beginning and end of response by splitting at first/last complete entry
    if "[{" not in resp:
        resp = "[{" + resp.split("{", 1)[1]
    if "}]" or "},]" not in resp:
        resp = resp.rpartition('}')[-3] + "}]"

    # remove potential text outside - no matter if it exists or not
    resp = ("[" + resp.split("[")[1]) 
    resp = (resp.split("]", 1)[0] + "]") 

    # save LLM response for each chunk  in dataframe for each document
    df_resp = pd.read_json(StringIO(resp))
    df_resp["citation_id"] = context[0]["citation"]  # add citation info
    # df_resp["chunk_id"] = j  # add chunk id as identifier
    # df_resp["ci_entity"] = df_ci_geo.loc[df_ci_geo["chunk_id"] == j, "ci_entity"] 
    # df_resp["geo_entity"] = df_ci_geo.loc[df_ci_geo["chunk_id"] == j, "geo_entity"]
    # df_resp["case_type"] = df_ci_geo.loc[df_ci_geo["chunk_id"] == j, "case_type"]
    df_resp["chunk_text"] =  chunk.page_content #  FIXME  context[0]["text"]  # add (translated) chunk text for tracing back LLM response

    df_responses = pd.concat([df_responses, df_resp], ignore_index=True)


except (IndexError, ValueError) as e:
    print(f"Cannot add response: {e},\nFaulty response (before postprocessing):", response[0]['generated_text'].replace('\n', ''))
    # faulty response: e.g.  .., "location": "V" on satellite and online on radar"}, { ...}, {}
    responses_error_list.append({
        "citation_id": citation,
        # "chunk_id": j,
        "response": response[0]['generated_text'].replace('\n', ''),
        "error": str(e)
    })

df_responses


In [ ]:
df_responses.loc[31, ["infrastructure_type", "damage", "location", "chunk_text"]]

# infrastructure_type:              bridge
# damage                            blocked
# location                          access road
# chunk_text             
#   At the peak of the event, around 200 000 people experienced power outages in Germany. Electricity infrastructure was severely damaged in North Rhine-Westphalia
#  and Rhineland-Palatinate. However, within 2 d around 50 % of the power was restored through repairs and temporary fixes. Within 8 weeks, no emergency power 
# generators were required any-more, with most of the power infrastructure restored in Ger-many’s affected areas. Some areas, however, only had perma-nent power 
# infrastructure after 6 months (Westnetz, 2022). The gas distribution network in the Ahr valley was severely damaged. Approximately 133 km of natural gas pipelines,
#  8500 gas metres, 3400 house pressure regulators, 7220 of the approximately 8000 household connections, and 31 systems measuring and regulating gas pressure have 
# been damaged or destroyed (SWR, 2021). Gas supply was almost fully re-stored within 4.5 months after the flood event (Energienetze Mittelrhein, 2021). 
# In Belgium, approximately 41 500 peo-ple experienced power outages at the peak of the event. This was the result of both damaged and deliberately switched-off 
# electrical cabinets to prevent serious damages. It took around 3 weeks to fully restore power.
# Name: 31, dtype: object

infrastructure_type                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     

In [33]:

gc.collect()
torch.cuda.empty_cache() 
torch.no_grad()
print(torch.cuda.memory_reserved() / 1e9)

8.95483904


In [ ]:
# ## moved out 
#    for chunk_no, chunk in enumerate(doc):

#         ## preprocess  TODO move to document cleaning workflow + dc.funcs
#         chunk.page_content = chunk.page_content.replace("\n", " ")
#         chunk.page_content = chunk.page_content.replace("- ", "-") # TODO test if ("- ", "") is better

#         # ## get most likely geolocation for each CI entity based on distance between tokens
#         # nlp_chunk = nlp(chunk.page_content)
#         # all_ents = [ent for ent in nlp_chunk.ents]
#         # ci_type_ents = [ent for ent in nlp_chunk.ents if ent.label_ in ["CI_TYPE", "FAC"]]

#         # # check if chunk contains CI_TYPE entities
#         # if len(ci_type_ents) > 0:

#         #     # iterate over all entities within chunk
#         #     for ent_idx in range(len(all_ents)):
#         #         # when entity is CI_TYPE or FAC (i.e. buidling, airports, highways) do following ...
#         #         if all_ents[ent_idx].label_ in ["CI_TYPE", "FAC"]:
#         #             ci_idx = ent_idx

#         #             ## .. calculate distances between CI_TYPE entity and  all GEO entities in chunk based on index position
#         #             distance_list = []
#         #             idx_in_chunk = []
#         #             try:
#         #                 for ent_idx in range(len(all_ents)):
#         #                     # TODO calc distances between CI_TYPE ~ GEO entities based on word numbers and not entities
#         #                     if all_ents[ent_idx].label_ in ["GPE", "LOC"]:
#         #                         geo_idx = ent_idx
#         #                         dist_ent_pair = np.abs(ci_idx - geo_idx)
#         #                         distance_list.append(dist_ent_pair)
#         #                         idx_in_chunk.append((ent_idx))
#         #                         closest_pair_idx = np.argmin(
#         #                             distance_list
#         #                         )  # idx of closest GEO entity
#         #                         distance_closest_pair = distance_list[closest_pair_idx]

#         #                 threshold = 5  # max token distance between CI_TYPE and GEO entity
#         #                 if distance_closest_pair > threshold:
#         #                     print(
#         #                         f""" Token distance is too large between CI_TYPE/FAR and next GEO entity which is of {distance_closest_pair} [token distance] > {threshold} [max. token distance] """
#         #                     )
#         #                     continue
#         #                 else:
#         #                     pass

#         #                 ## write as dict entry incl chunk_id, ci_entity, geo_entity, distance
#         #                 result_dict = {
#         #                     "chunk_id": chunk_no,
#         #                     "ci_entity": all_ents[ci_idx].text,
#         #                     "ci_entity_label": all_ents[ci_idx].label_,
#         #                     "geo_entity": all_ents[idx_in_chunk[closest_pair_idx]].text,
#         #                     "geo_entity_label": all_ents[idx_in_chunk[closest_pair_idx]].label_,
#         #                     "token_distance": distance_closest_pair,
#         #                 }
#         #                 df_ci_geo = pd.concat(
#         #                     [df_ci_geo, pd.DataFrame([result_dict])], ignore_index=True
#         #                 )

#         #             except IndexError:
#         #                 continue
#         # else:
#         #     # print("\nNo CI_TYPE or FAC entities found in this chunk.")
#         #     continue


#     ## post-process of DF CI-GEO pairs
#     print(
#         "number of duplicates:", len(df_ci_geo.drop_duplicates(
#             subset=["citation_id", "chunk_id", "ci_entity", "geo_entity","case_type", "chunk_text"]))
#     )
#     df_ci_geo = df_ci_geo.drop_duplicates(
#         subset=["citation_id", "chunk_id", "ci_entity", "geo_entity","case_type", "chunk_text"]
#         )# .reset_index(drop=True, inplace=True)

############################
# ...
###########################

        # ]
        # # else:  # TODO dissolve if else clause by making it in ci_locations: if df_ci_geo.chunk=j, xx, else None
        # #     context = [
        # #                 {
        # #                     "text": chunk.page_content,
        # #                     "citation": citation,
        # #                     "title": filename_stem, 
        # #                     "ci_locations": df_ci_geo.loc[df_ci_geo["chunk_id"] == j],
        # #                 },
        # #             ]

        # # load prompt template
        # template = load_prompt_template(template_filename="ci_loc_direct_impacts.txt")


In [ ]:
response

In [ ]:
# df_ci_geo
j = 3
df_resp["ci_entity"] = df_ci_geo.loc[df_ci_geo["chunk_id"] == j, "ci_entity"]
df_resp.sort_values(by="ci_entity")

In [ ]:
# 12 docs with fixed NERpatterns: 74min

In [ ]:
print(torch.cuda.memory_reserved() / 1e9)


In [ ]:
print("Chunk with erroneous responses:", responses_error_list.__len__())
df_responses_error = pd.DataFrame(responses_error_list)
df_responses_error#.tail(3)


In [ ]:
df_responses_all#.tail(3)

### Saving

In [ ]:
# PATH_LLM_DATA: Path = Path(s.PATH_DATA /"llm_outputs/")
# LLM_DATA_FILENAME: str = "llm_1_updprompt_distanceNER.csv"

# OUTPUT_LLM1_FILEPATH =  Path(PATH_LLM_DATA / LLM_DATA_FILENAME)#.replace(".csv", "_v2.csv"))
# OUTPUT_LLM1_FILEPATH 
df_responses_all.to_csv("geollm_1st.csv", index=False)


In [ ]:
safety_df = df_responses_all.copy()



# save LLM 1 output to disk along with prompt text
if not os.path.isfile(OUTPUT_LLM1_FILEPATH):

    print(f"Saving prompt, LLM response and erroneous responses [.txt, .csv] to {OUTPUT_LLM1_FILEPATH} ...")

    with open(OUTPUT_LLM1_FILEPATH.parent / f"prompt_{OUTPUT_LLM1_FILEPATH.stem}.txt", "w") as f:
        f.write(template.render(context=context, question=question_1))
    df_responses_all.to_csv(OUTPUT_LLM1_FILEPATH, index=False)
    df_responses_error.to_csv(OUTPUT_LLM1_FILEPATH.parent / f"errors_{OUTPUT_LLM1_FILEPATH.stem}.csv", index=False)

elif os.path.isfile(OUTPUT_LLM1_FILEPATH) and not os.path.isfile(OUTPUT_LLM1_FILEPATH.parent / f"{OUTPUT_LLM1_FILEPATH.stem}_v2.csv"):

    print(f"Output file {Path(OUTPUT_LLM1_FILEPATH).stem} already exists. Saving as {OUTPUT_LLM1_FILEPATH.stem}_v2 to avoid overwriting ...")

    # If the original files exists but the v2 file doesn't, create the v2 file
    with open(OUTPUT_LLM1_FILEPATH.parent / f"prompt_{OUTPUT_LLM1_FILEPATH.stem}_v2.txt", "w") as f:
        f.write(template.render(context=context, question=question_1))
    df_responses_all.to_csv(Path(OUTPUT_LLM1_FILEPATH.parent, f"{OUTPUT_LLM1_FILEPATH.stem}_v2.csv"), index=False)
    df_responses_error.to_csv(Path(OUTPUT_LLM1_FILEPATH.parent / f"errors_{OUTPUT_LLM1_FILEPATH.stem}_v2.csv"), index=False)

else:
    print(f"Output file {Path(OUTPUT_LLM1_FILEPATH).stem} already exists. Skip saving to avoid overwriting ...")






Save as pyarrow incl dtypes as pyarraws-. Saving also dtypes a spyarrow objectes leads to quicker loading and processing

In [ ]:
# pd.options.future.infer_string = True

df_responses_all_pyarrow = df_responses_all.astype( dtype="string[pyarrow]")
df_responses_all_pyarrow = pa.Table.from_pandas(df_responses_all_pyarrow)
pq.write_table(df_responses_all_pyarrow, OUTPUT_LLM1_FILEPATH.with_suffix(".parquet"))  



## Evaluation - only manually

In [ ]:
df_responses_all.sort_values(by=["ci_entity"])[:50]  # == "Spain’s second-biggest port"]

### Manual comparison CI_location_table vs LLm response


#### chunk 5
In Germany, road and railway infrastructure was severely damaged as documented exemplarily in Fig. 1. Cost estimates reach up to EURO 2 billion Euro (MDR, 2021). More than 130 km of motorways were closed directly after the event, of which 50 km were still closed two months later, with an estimated repair cost of EUR 100 million (Hauser, 2021). Of the 112 bridges in the flooded 40 km of the Ahr valley (Rhineland-Palatinate), 62 bridges were destroyed, 13 were severely damaged and only 35 were in operation a month after the flood event (MDR, 2021). Over 74 km of roads, paths and bridges in the Ahr valley have been (critically) damaged. In some cases, repairs are expected to take months to years (Zeit Online, 2021). For example, major freeway sections, including parts of the A1 motorway, were closed until early 2022 (24Rhein, 2022). In addition, about 50 000 cars were damaged, causing insurance claims of some EUR 450 million (ADAC, 2021). The German railway provider Deutsche Bahn expects asset damages of around EUR 1.3 billion. Among other things, 180 level crossings, almost 40 signal'

In [ ]:
doc[5].page_content

In [ ]:
df_ci_geo[df_ci_geo["chunk_id"] == 5]

In [ ]:
df_responses[df_responses["chunk_id"] == 5]

In [ ]:
print(
    f"CI_TYPE \n LLM responses:\n {df_responses.infrastructure_type.unique()}, \n\n NER pairs:\n {df_ci_geo.ci_entity.unique()}"
)

In [ ]:
print(
    f"LOCATION:\n LLm responses:\n {df_responses.location.unique()}, \n\n NER pairs:\n {df_ci_geo.geo_entity.unique()}"
)

### chunk 13 +14
'We found no information regarding direct impact on solid-waste facilities as a result of the flood event. However, there is a large pressure on the solid-waste sector to clean the affected areas; 1 month after the event, we observed dozens of large temporary waste fills and frequent incidences of oil pollution in Rhineland-Palatinate during a field visit. In the Ahrweiler district alone, the flood caused as much solid waste as normally would be collected over 30 years. In Belgium, the amount of solid waste is estimated around 160 000 t, stored at several places, such as the abandoned highway track A601. This highway has been used for approximately 9 months as a temporary storage for debris (Couplez, 2022). In the Netherlands, there have been primarily problems with waste deposits along the river banks, which is mostly the solid waste transported by the river from further upstream. Thousands of tonnes of tree debris (logs and\nNat. Hazards Earth Syst. Sci., 22, 3831–3838, 2022\nE. E. Koks et al.: Flood impacts to infrastructure'


'of running water and electricity (Ärzte Zeitung, 2021). After 1.5 months, medical care was guaranteed again in the most affected regions in Rhineland-Palatinate (Hochwasser Ahr, 2021c). In the state of North Rhine-Westphalia, approximately 68 hospitals have been affected, of which several have been affected severely and will take at least 1.5 years to be rebuilt (Fig. 2). Direct damages are estimated to be at least EUR 100 million to repair all medical facilities (Korzilius, 2021). In the town of Eschweiler (Germany), for example, the basement of the hospital was flooded, as well as the outbuildings and the entire outdoor area. The power supply collapsed, the entire building technology was destroyed and some 300 patients had to be evacuated by helicopter. Property damage is expected to be around EUR 50 million. Within 3.5 weeks, the hospital was partly operational, and within 3 months, all hospital operations continued normally (SAH Eschweiler, 2021). The Mutterhaus Ehrang hospital in Trier (Germany) is now permanently closed as the hospital is too severely damaged to rebuild. Furthermore, in the region of Rhineland-Palatinate (Germany), 19'


In [ ]:
doc[14].page_content

In [ ]:
df_ci_geo[df_ci_geo["chunk_id"].isin([13, 14])]

In [ ]:
df_responses[df_responses["chunk_id"].isin([13, 14])]

In [ ]:
print(
    f"CI_TYPE \n LLM responses:\n {df_responses.infrastructure_type.unique()}, \n\n NER pairs:\n {df_ci_geo.ci_entity.unique()}"
)

In [ ]:
print(
    f"LOCATION:\n LLm responses:\n {df_responses.location.unique()}, \n\n NER pairs:\n {df_ci_geo.geo_entity.unique()}"
)

## Test alternative approaches for entity linking / relation extraction

In [ ]:
print(torch.cuda.memory_summary(device=None, abbreviated=False))
# # empyty CUDA cache
gc.collect()

torch.cuda.empty_cache()
torch.no_grad()
# print(torch.cuda.memory_summary(device=None, abbreviated=False))